### PR TITLE
Even more `ObligationForest` improvements

### PR DESCRIPTION
**Pull request by nnethercote**
Following on from #64545, more speed and readability improvements.

r? @nikomatsakis

### CODE REVIEW

**Comment by nnethercote** [comment]
Not sure what happened there. Let's try it the old-fashioned way:
@bors try

**Comment by nnethercote** [comment]
Some local measurements:
```
inflate-check
        avg: -10.3%     min: -14.3%     max: 0.1%
keccak-check
        avg: -7.8%      min: -12.0%     max: 0.0%
cranelift-codegen-check
        avg: -1.2%      min: -2.0%      max: 0.0%
clap-rs-check
        avg: -0.6%      min: -1.6%      max: -0.0%
serde-check
        avg: -0.5%      min: -0.8%      max: 0.1%
unify-linearly-check
        avg: -0.3%      min: -0.7%      max: 0.0% 
deeply-nested-check
        avg: -0.4%      min: -0.6%      max: 0.0% 
style-servo-check
        avg: -0.4%      min: -0.6%      max: 0.0%
wg-grammar-check
        avg: -0.3%      min: -0.5%      max: -0.0%
cargo-check
        avg: -0.2%      min: -0.3%      max: 0.0%
syn-check
        avg: -0.1%      min: -0.3%      max: 0.0% 
```

**Comment by bjorn3** [comment]
@nnethercote Do you every take a day off of optimizing rustc? :smile:

File: src/librustc/traits/fulfill.rs [hunk 1]
```diff
@@ -256,29 +256,46 @@ impl<'a, 'b, 'tcx> ObligationProcessor for FulfillProcessor<'a, 'b, 'tcx> {
...
-            if !changed {
-                debug!("process_predicate: pending obligation {:?} still stalled on {:?}",
-                       self.selcx.infcx()
-                           .resolve_vars_if_possible(&pending_obligation.obligation),
-                       pending_obligation.stalled_on);
-                return ProcessResult::Unchanged;
+            _ => {
+                // This `for` loop was once a call to `all()`, but this lower-level
+                // form was a perf win. See #64545 for details.
+                (|| {
```
Comment (nikomatsakis) [inline src/librustc/traits/fulfill.rs]: Heh, cute. I presume the intention here is to "outline" the cold part?

**Review by nikomatsakis — APPROVED** [review]

File: src/librustc/traits/fulfill.rs (hunk 1 above)
Comment (nnethercote) [reply src/librustc/traits/fulfill.rs]: I initially wrote it as a normal loop that set a boolean flag and then did `break`, but this version was more concise (due to being able to use `return`) and measurements showed it was also slightly faster (fewer instructions). This code is very perf-sensitive for benchmarks like `keccak` and `inflate` and seemingly trivial differences often result in measurable changes.

**Pull request merged**